Add TLS certificate auth for HashiCorp Vault

marbindrakon opened the pull request:
##### SUMMARY
Add support for AWX to authenticate with HashiCorp Vault using TLS client certificates.

Also updates the documentation for the HashiCorp Vault secret management plugins to include both the new TLS options and the missing Kubernetes auth method options.

##### ISSUE TYPE
 - New or Enhanced Feature

##### COMPONENT NAME
HashiCorp Vault Secret Management

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 0.1.dev33561+g6b06f8d

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
Some users prefer to use TLS client certificates for authentication with HashiCorp Vault rather than AppRole, Token, or Kubernetes authentication. This PR adds fields to the built-in credential types for HashiCorp Vault to support client certificate authentication as well as the logic to use that certificate to obtain the session Vault token. 

The current CertFiles implementation is used to handle the client certificate data in temporary files as required by the requests library.

fosterseth commented:
really nice contribution! 

We'd love to see our bootstrapped vault server be setup with TLS

https://github.com/ansible/awx/blob/devel/tools/docker-compose/ansible/roles/vault/tasks/initialize.yml

This will help us test this feature out.

would you mind updating that playbook to deploy vault with TLS enabled as part of this PR?

Then setup a vault credential in AWX to connect to this TLS enabled vault https://github.com/ansible/awx/blob/devel/tools/docker-compose/ansible/roles/vault/tasks/plumb.yml

marbindrakon commented:
> really nice contribution!
> 
> We'd love to see our bootstrapped vault server be setup with TLS
> 
> https://github.com/ansible/awx/blob/devel/tools/docker-compose/ansible/roles/vault/tasks/initialize.yml
> 
> This will help us test this feature out.
> 
> would you mind updating that playbook to deploy vault with TLS enabled as part of this PR?
> 
> Then setup a vault credential in AWX to connect to this TLS enabled vault https://github.com/ansible/awx/blob/devel/tools/docker-compose/ansible/roles/vault/tasks/plumb.yml

Sure thing. I should be able to make it work with some self-signed certs.

TheRealHaoLiu commented:
```
linters: commands[1]> flake8 awx awxkit awx_collection
awx/main/tests/functional/test_credential_plugins.py:102:5: F841 local variable 'auth_params' is assigned to but never used
linters: exit 1 (15.22 seconds) /awx_devel> flake8 awx awxkit awx_collection pid=284
.pkg: _exit> python /var/lib/awx/venv/awx/lib64/python3.9/site-packages/pyproject_api/_backend.py True setuptools.build_meta
  linters: FAIL code 1 (64.89=setup[34.55]+cmd[15.12,15.22] seconds)
  evaluation failed :( (65.34 seconds)
make: *** [Makefile:329: docker-runner] Error 1
```

```
=================================== FAILURES ===================================
___________________ test_hashivault_handle_auth_client_cert ____________________
[gw0] linux -- Python 3.9.17 /var/lib/awx/venv/awx/bin/python3.9
Traceback (most recent call last):
  File "/awx_devel/awx/main/tests/functional/test_credential_plugins.py", line 107, in test_hashivault_handle_auth_client_cert
    token = hashivault.handle_auth(**kwargs)
  File "/awx_devel/awx/main/credential_plugins/hashivault.py", line 198, in handle_auth
    raise Exception('Either a token or AppRole, Kubernetes, or TLS authentication parameters must be set')
Exception: Either a token or AppRole, Kubernetes, or TLS authentication parameters must be set
```

there's some CI failures

marbindrakon committed:
Add TLS certificate auth for HashiCorp Vault

Add support for AWX to authenticate with HashiCorp Vault using
TLS client certificates.

Also updates the documentation for the HashiCorp Vault secret management
plugins to include both the new TLS options and the missing Kubernetes
auth method options.

Signed-off-by: Andrew Austin <aaustin@redhat.com>

marbindrakon committed:
Refactor docker-compose vault for TLS cert auth

Add TLS configuration to the docker-compose Vault configuration and
use that method by default in vault plumbing.

This ensures that the result of bringing up the docker-compose stack
with vault enabled and running the plumb-vault playbook is a fully
working credential retrieval setup using TLS client cert authentication.

Signed-off-by: Andrew Austin <aaustin@redhat.com>

marbindrakon commented:
@fosterseth It took a bit more effort than I planned, but the docker-compose setup now bootstraps TLS and the plumb_vault playbook configures a working credential source using TLS auth.

marbindrakon commented:
This should be ready to re-run CI tests on if that can be approved.

TheRealHaoLiu commented:
@marbindrakon thank you! i will pull down the code and try it out!

TheRealHaoLiu commented on tools/docker-compose/ansible/roles/sources/defaults/main.yml:
```diff
@@ -30,6 +30,24 @@ ldap_public_key_file: '{{ ldap_cert_dir }}/{{ ldap_public_key_file_name }}'
+hashivault_server_private_keyfile: '{{ hashivault_cert_dir }}/server.key'
+hashivault_client_public_keyfile: '{{ hashivault_cert_dir }}/client.crt'
+hashivault_client_private_keyfile: '{{ hashivault_cert_dir }}/client.key'
+  
```
```suggestion

```

marbindrakon committed:
Remove incorrect trailing space

Co-authored-by: Hao Liu <44379968+TheRealHaoLiu@users.noreply.github.com>

TheRealHaoLiu commented:
oops accidentally push a PR into the your branch... didn't know i have the permission...

marbindrakon commented:
> oops accidentally push a PR into the your branch... didn't know i have the permission...

Moved the history of my branch back to before your PR commit. That timeline is kept in the rebase-backup branch on my fork if you were using it.

TheRealHaoLiu commented:
> > oops accidentally push a PR into the your branch... didn't know i have the permission...
> 
> Moved the history of my branch back to before your PR commit. That timeline is kept in the rebase-backup branch on my fork if you were using it.

thanks! i'm continuing for the review right now. i think we probably want to put vault tls behind a flag

TheRealHaoLiu committed:
Make vault init idempotent

- improve error handling for vault_initialization
- ignore error if vault cert auth is already configured
- removed unused register

TheRealHaoLiu committed:
Add VAULT_TLS option

Make TLS for HashiCorp Vault optional and configurable via VAULT_TLS env var

TheRealHaoLiu commented:
@marbindrakon 

I can't create PR to your fork for some reason... (but somehow i have permission to push to your fork?)

so I open a PR with some modification to your code https://github.com/ansible/awx/pull/14691

mainly this commit https://github.com/ansible/awx/pull/14691/commits/f2f7d18cf9c74c6ea68b8fcb50b770ac776ae67d 

can you take a look

marbindrakon commented:
Thanks for your improvements! The only comment I have is that plumb.yml will not function without TLS enabled since I changed it to use the TLS client auth. It was completely broken before this PR as far as I can tell, so I don't know if we need to gold plate it by having it use token auth or something if vault_tls is fasle.

marbindrakon commented:
Do we want to close this PR and work in your branch @TheRealHaoLiu ?

TheRealHaoLiu commented:
If ur ok with it i can commit my change to your branch. I made some change for plumb to detect if the cert directory exist and use http/https

TheRealHaoLiu commented:
If ur ok with it i can commit my change to your branch. I made some change for plumb to detect if the cert directory exist and use http/https

marbindrakon commented:
@TheRealHaoLiu Go ahead

TheRealHaoLiu committed:
Add retries for vault init

Sometime it took longer for vault to fully come up and init will fail

TheRealHaoLiu commented:
i tested and made sure that when VAULT_TLS is unset or false plumb_vault will succeed
